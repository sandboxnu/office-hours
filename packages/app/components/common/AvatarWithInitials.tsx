import { Avatar } from "antd";
import React, { ReactElement, Ref } from "react";
import nameToRGB from "../../utils/ColorUtils";
import getInitialsFromName from "../../utils/NameUtils";

type AvatarWithInitalsProps = {
  name: string;
  fontSize?: number;
  style?: any;
  [x: string]: any;
};

export const AvatarWithInitals = React.forwardRef(
  (
    { name, fontSize, style, ...props }: AvatarWithInitalsProps,
    ref: Ref<HTMLElement>
  ): ReactElement => {
    return (
      <Avatar
        {...props}
        ref={ref}
        style={{
          backgroundColor: name ? nameToRGB(name) : "#1abc9c",
          fontSize,
          ...style,
        }}
      >
        {getInitialsFromName(name)}
      </Avatar>
    );
  }
);

AvatarWithInitals.displayName = "AvatarWithInitials";
