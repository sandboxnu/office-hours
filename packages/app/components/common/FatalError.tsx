import { apm } from "@elastic/apm-rum";
import Result from "antd/lib/result";
import { ReactElement, useEffect } from "react";

export function FatalError({ error }: { error: Error }): ReactElement {
  useEffect(() => window && apm.captureError(error), [error]);
  return (
    <Result
      status="500"
      title="We hit an unexpected error."
      subTitle="Sorry about that! A report has automatically been filed. Try refreshing the page."
    />
  );
}
