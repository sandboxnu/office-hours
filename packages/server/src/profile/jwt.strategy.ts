import { Strategy } from 'passport-jwt';
import { PassportStrategy } from '@nestjs/passport';
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Request } from 'express';

@Injectable()
export class JwtStrategy extends PassportStrategy(Strategy) {
  constructor(configService: ConfigService) {
    super({
      jwtFromRequest: (req: Request) => req.cookies['auth_token'],
      ignoreExpiration: false,
      secretOrKey: configService.get('JWT_SECRET'),
    });
  }

  validate(payload: { userId: number }): any {
    return { ...payload };
  }
}
