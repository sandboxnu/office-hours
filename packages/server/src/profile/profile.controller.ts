import { Body, Controller, Get, Patch, UseGuards } from '@nestjs/common';
import { GetProfileResponse, UpdateProfileParams } from '@template/common';
import { pick } from 'lodash';
import { PhoneNotifModel } from 'notification/phone-notif.entity';
import { Connection } from 'typeorm';
import { JwtAuthGuard } from '../login/jwt-auth.guard';
import { NotificationService } from '../notification/notification.service';
import { User } from './user.decorator';
import { UserModel } from './user.entity';

@Controller('profile')
@UseGuards(JwtAuthGuard)
export class ProfileController {
  constructor(
    private connection: Connection,
    private notifService: NotificationService,
  ) {}

  @Get()
  async get(
    @User(['courses', 'courses.course']) user: UserModel,
  ): Promise<GetProfileResponse> {
    const courses = user.courses
      .filter((userCourse) => userCourse.course.enabled)
      .map((userCourse) => {
        return {
          course: {
            id: userCourse.courseId,
            name: userCourse.course.name,
          },
          role: userCourse.role,
        };
      });

    const phoneNotifModel = await PhoneNotifModel.findOne({
      where: {
        userId: user.id,
      },
    });

    const userResponse = pick(user, [
      'id',
      'email',
      'name',
      'photoURL',
      'desktopNotifsEnabled',
      'phoneNotifsEnabled',
    ]);
    return {
      ...userResponse,
      courses,
      phoneNumber: phoneNotifModel?.phoneNumber,
    };
  }

  @Patch()
  async patch(
    @Body() userPatch: UpdateProfileParams,
    @User(['courses', 'courses.course']) user: UserModel,
  ): Promise<GetProfileResponse> {
    user = Object.assign(user, userPatch);
    if (user.phoneNotifsEnabled && userPatch.phoneNumber) {
      await this.notifService.registerPhone(userPatch.phoneNumber, user.id);
    }
    await user.save();

    const courses = user.courses.map((userCourse) => {
      return {
        course: {
          id: userCourse.courseId,
          name: userCourse.course.name,
        },
        role: userCourse.role,
      };
    });

    const phoneNotifModel = await PhoneNotifModel.findOne({
      where: {
        userId: user.id,
      },
    });

    const userResponse = pick(user, [
      'id',
      'email',
      'name',
      'photoURL',
      'desktopNotifsEnabled',
      'phoneNotifsEnabled',
    ]);
    return {
      ...userResponse,
      courses,
      phoneNumber: phoneNotifModel.phoneNumber,
    };
  }
}
