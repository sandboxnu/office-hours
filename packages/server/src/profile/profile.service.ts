import { KhouryProfCourse, QuestionStatusKeys } from '@koh/common';
import { Injectable } from '@nestjs/common';
import {
  LastRegistrationModel,
  khourySemesterCodes,
} from 'login/last-registration-model.entity';
import { LoginCourseService } from 'login/login-course.service';
import { ProfSectionGroupsModel } from 'login/prof-section-groups.entity';
import { Connection } from 'typeorm';
import { UserModel } from './user.entity';
import { QuestionModel } from 'question/question.entity';

@Injectable()
export class ProfileService {
  constructor(
    private connection: Connection,
    private loginCourseService: LoginCourseService,
  ) {}

  public async inQueue(user: UserModel): Promise<boolean> {
    const question = await QuestionModel.findOne({
      where: {
        creatorId: user.id,
      },
    });
    if (question) {
      if (question.status === QuestionStatusKeys.Queued) {
        return true;
      }
      return false;
    }
    return false;
  }
  public async getPendingCourses(userId: number): Promise<KhouryProfCourse[]> {
    const profCourses = await ProfSectionGroupsModel.findOne({
      where: {
        profId: userId,
      },
    });
    const lastRegistered = await LastRegistrationModel.findOne({
      where: {
        profId: userId,
      },
    });
    const lastRegisteredSemester = lastRegistered?.lastRegisteredSemester;

    if (!profCourses) return; // not a professor

    const pendingCourses = [];

    for (const c of profCourses.sectionGroups) {
      // current semester does not match last time prof registered, there may be pending courses
      if (
        c.crns.length !== 0 &&
        (!lastRegisteredSemester || // lastRegistered doesnt exist if prof has never registered before
          !this.isSameRegistrationTime(c.semester, lastRegisteredSemester))
      ) {
        const courseCRN = c.crns[0];
        const profCourse = await this.loginCourseService.courseCRNToCourse(
          courseCRN,
          c.semester,
        );
        if (!profCourse) pendingCourses.push(c);
      }
    }

    return pendingCourses;
  }

  // hacky way to check if two semester codes from Khoury have the same registration window
  private isSameRegistrationTime(semester1: string, semester2: string) {
    const year1 = Number(semester1.slice(0, 4));
    const year2 = Number(semester2.slice(0, 4));
    const semesterCode1 = semester1.slice(-2);
    const semesterCode2 = semester2.slice(-2);
    const summer1 = khourySemesterCodes['Summer_1'];
    const summerFull = khourySemesterCodes['Summer_Full'];
    // we want to treat Summer 1 and Summer Full as the same window
    // for profs to register classes for both at the same time
    if (
      (semesterCode1 === summer1 || semesterCode2 === summer1) &&
      (semesterCode1 === summerFull || semesterCode2 === summerFull)
    ) {
      return year1 === year2;
    }
    return semester1 === semester2;
  }
}
