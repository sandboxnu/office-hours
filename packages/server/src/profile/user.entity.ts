import { Exclude } from 'class-transformer';
import { INSIGHTS_MAP } from '../insights/insight-objects';
import * as bcrypt from 'bcrypt';
import {
  AfterLoad,
  BaseEntity,
  BeforeInsert,
  Column,
  Entity,
  JoinColumn,
  ManyToMany,
  OneToMany,
  OneToOne,
  PrimaryGeneratedColumn,
} from 'typeorm';
import { DesktopNotifModel } from '../notification/desktop-notif.entity';
import { PhoneNotifModel } from '../notification/phone-notif.entity';
import { QueueModel } from '../queue/queue.entity';
import { EventModel } from './event-model.entity';
import { UserCourseModel } from './user-course.entity';
import { AlertModel } from '../alerts/alerts.entity';
import { UserRole } from '@koh/common';
import { OrganizationUserModel } from '../organization/organization-user.entity';
import { InteractionModel } from 'chatbot/interaction.entity';

@Entity('user_model')
export class UserModel extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;
  @Column('int', { nullable: true })
  sid: number;

  @Column('text')
  email: string;

  @Column('text', { nullable: true })
  password: string | null;

  @Column('text', { nullable: true })
  firstName: string;

  @Column('text', { nullable: true })
  lastName: string;

  @Column('text', { nullable: true })
  photoURL: string | null;

  @Column('text', { nullable: true })
  defaultMessage: string | null;

  @Column({ type: 'boolean', default: true })
  includeDefaultMessage: boolean;

  @Column({ type: 'boolean', default: false })
  accountDeactivated: boolean;

  @OneToMany((type) => UserCourseModel, (ucm) => ucm.user)
  @Exclude()
  courses: UserCourseModel[];

  @Column({ type: 'boolean', default: false })
  @Exclude()
  desktopNotifsEnabled: boolean; // Does user want notifications sent to their desktops?

  @Column({ type: 'boolean', default: false })
  @Exclude()
  phoneNotifsEnabled: boolean; // Does user want notifications sent to their phone?

  @Column({ type: 'enum', enum: UserRole, default: UserRole.USER })
  userRole: UserRole;

  @OneToMany((type) => DesktopNotifModel, (notif) => notif.user)
  @Exclude()
  desktopNotifs: DesktopNotifModel[];

  @OneToOne((type) => PhoneNotifModel, (notif) => notif.user)
  @Exclude()
  phoneNotif: PhoneNotifModel;

  @Exclude()
  @ManyToMany((type) => QueueModel, (queue) => queue.staffList)
  queues: QueueModel[];

  @Exclude()
  @OneToMany((type) => EventModel, (event) => event.user)
  events: EventModel[];

  @OneToMany((type) => AlertModel, (alert) => alert.user)
  alerts: AlertModel[];

  @Exclude()
  @Column({ type: 'simple-array', nullable: true })
  hideInsights: string[];

  insights: string[];

  @OneToOne((type) => OrganizationUserModel, (ou) => ou.organizationUser)
  organizationUser: OrganizationUserModel;

  @OneToMany((type) => InteractionModel, (interaction) => interaction.user)
  @JoinColumn({ name: 'user' })
  interactions: InteractionModel[];

  @AfterLoad()
  computeInsights(): void {
    let hideInsights = this.hideInsights;
    if (!hideInsights) {
      hideInsights = [];
    }
    const insightNames = Object.keys(INSIGHTS_MAP);
    this.insights = insightNames.filter((name) => !hideInsights.includes(name));
  }

  name: string;

  @AfterLoad()
  setFullNames(): void {
    this.name = this.firstName + ' ' + this.lastName;
  }
  // @BeforeInsert()
  // async setPassword(password:string) {
  //   const saltRounds = 10;
  //   const salt = await bcrypt.genSalt(saltRounds);
  //   this.password = await bcrypt.hash(password||this.password, salt);
  // }
}
