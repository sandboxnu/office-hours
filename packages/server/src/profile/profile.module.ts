import { Module } from '@nestjs/common';
import { ProfileController } from './profile.controller';
import { NotificationModule } from '../notification/notification.module';
import { ProfileService } from './profile.service';
import { LoginModule } from 'login/login.module';
import { JwtModule } from '@nestjs/jwt';
import { LoginCourseService } from 'login/login-course.service';
import { ConfigModule, ConfigService } from '@nestjs/config';
import { JwtStrategy } from './jwt.strategy';
import { MailModule } from 'mail/mail.module';

@Module({
  imports: [
    NotificationModule,
    LoginModule,
    MailModule,
    JwtModule.registerAsync({
      imports: [ConfigModule],
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) => ({
        secret: configService.get('JWT_SECRET'),
      }),
    }),
  ],
  controllers: [ProfileController],
  providers: [JwtStrategy, ProfileService, LoginCourseService],
})
export class ProfileModule {}
