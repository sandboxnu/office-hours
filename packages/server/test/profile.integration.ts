import {
  StudentCourseFactory,
  UserFactory,
  CourseFactory,
  ProfSectionGroupsFactory,
  CourseSectionFactory,
  LastRegistrationFactory,
  OrganizationFactory,
} from './util/factories';
import { setupIntegrationTest } from './util/testUtils';
import { ProfileModule } from '../src/profile/profile.module';
import { PhoneNotifModel } from 'notification/phone-notif.entity';
import { DesktopNotifModel } from 'notification/desktop-notif.entity';
import { OrganizationUserModel } from 'organization/organization-user.entity';
import { AccountType } from '@koh/common';

describe('Profile Integration', () => {
  const supertest = setupIntegrationTest(ProfileModule);

  describe('GET /profile', () => {
    it('returns the logged-in user profile', async () => {
      const organization = await OrganizationFactory.create();
      const user = await UserFactory.create();
      const fundies = await CourseFactory.create({ name: 'CS 2500' });
      await StudentCourseFactory.create({ course: fundies, user });
      await OrganizationUserModel.create({
        userId: user.id,
        organizationId: organization.id,
      }).save();

      const res = await supertest({ userId: user.id })
        .get('/profile')
        .expect(200);
      expect(res.body).toMatchSnapshot();
    });

    it('returns only userCourses where course is enabled', async () => {
      const user = await UserFactory.create();
      const fundies = await CourseFactory.create({ name: 'CS 2500' });
      const nonEnabledCourse = await CourseFactory.create({
        name: 'CS 4900',
        enabled: false,
      });
      await StudentCourseFactory.create({ course: fundies, user });
      await StudentCourseFactory.create({ course: nonEnabledCourse, user });

      const res = await supertest({ userId: user.id })
        .get('/profile')
        .expect(200);

      expect(res.body.courses).toEqual([
        {
          course: {
            id: 1,
            name: 'CS 2500',
          },
          role: 'student',
        },
      ]);
    });

    it('returns desktop notif information', async () => {
      const user = await UserFactory.create();
      const dn = await DesktopNotifModel.create({
        user,
        auth: '',
        p256dh: '',
        endpoint: 'abc',
        name: 'firefox',
      }).save();
      await dn.reload();
      const res = await supertest({ userId: user.id })
        .get('/profile')
        .expect(200);
      expect(res.body.desktopNotifs).toEqual([
        {
          createdAt: expect.any(String),
          name: 'firefox',
          id: dn.id,
          endpoint: dn.endpoint,
        },
      ]);
    });

    it('returns 401 when not logged in', async () => {
      await UserFactory.create();
      await supertest().get('/profile').expect(401);
    });

    it('returns pending courses when they exist', async () => {
      const prof1KhouryCourses = [
        {
          name: 'Fundies 1',
          crns: [123, 456],
          semester: '202110',
        },
        {
          name: 'OOD',
          crns: [798],
          semester: '202110',
        },
      ];
      const prof1 = await UserFactory.create();
      await ProfSectionGroupsFactory.create({
        prof: prof1,
        sectionGroups: prof1KhouryCourses,
      });
      await LastRegistrationFactory.create({
        prof: prof1,
        lastRegisteredSemester: '202010',
      });
      const fundies1 = await CourseFactory.create();
      await CourseSectionFactory.create({ crn: 123, course: fundies1 });
      await CourseSectionFactory.create({ crn: 456, course: fundies1 });

      const res = await supertest({ userId: prof1.id })
        .get('/profile')
        .expect(200);

      expect(res.body.pendingCourses).toEqual([
        prof1KhouryCourses[0],
        prof1KhouryCourses[1],
        // {
        //   name: 'OOD',
        //   crns: [798],
        //   semester: '202110',
        // },
      ]);
    });
  });

  describe('DELETE /profile/delete_profile_picture', () => {
    it("should delete profile picture from database only if it's URL", async () => {
      const user = await UserFactory.create({
        photoURL: 'https://www.google.com',
      });
      await supertest({ userId: user.id })
        .delete('/profile/delete_profile_picture')
        .expect(200);
    });
  });

  describe('PATCH /profile', () => {
    it('enables desktop notifs', async () => {
      const user = await UserFactory.create({
        desktopNotifsEnabled: false,
        phoneNotifsEnabled: false,
      });
      const res = await supertest({ userId: user.id })
        .patch('/profile')
        .send({ desktopNotifsEnabled: true })
        .expect(200);
      expect(res.body).toMatchObject({
        message: 'Profile updated successfully',
      });
    });

    it('enables phone notifs', async () => {
      const user = await UserFactory.create({
        desktopNotifsEnabled: false,
        phoneNotifsEnabled: false,
      });
      const res = await supertest({ userId: user.id })
        .patch('/profile')
        .send({ phoneNotifsEnabled: true, phoneNumber: '911' })
        .expect(200);
      expect(res.body).toMatchObject({
        message: 'Profile updated successfully',
      });
    });

    it('should not update profile settings if email is included and account is not legacy', async () => {
      const user = await UserFactory.create({
        email: 'test@ubc.ca',
        accountType: AccountType.GOOGLE,
      });

      const res = await supertest({ userId: user.id })
        .patch('/profile')
        .send({ email: 'test@ubc.ca', firstName: 'test' });

      expect(res.status).toEqual(400);
      expect(res.body).toEqual({
        message: 'Email cannot be updated',
      });
    });

    it('does not let student enable without phone number', async () => {
      const user = await UserFactory.create({
        desktopNotifsEnabled: false,
        phoneNotifsEnabled: false,
      });
      await supertest({ userId: user.id })
        .patch('/profile')
        .send({ phoneNotifsEnabled: true })
        .expect(400);
    });

    it('lets student change phone number', async () => {
      const user = await UserFactory.create({
        desktopNotifsEnabled: false,
        phoneNotifsEnabled: true,
      });
      await PhoneNotifModel.create({
        phoneNumber: '1234567890',
        user: user,
        verified: true,
      }).save();
      let profile = await supertest({ userId: user.id }).get('/profile');
      expect(profile.body?.phoneNumber).toEqual('1234567890');
      await supertest({ userId: user.id })
        .patch('/profile')
        .send({ phoneNotifsEnabled: true, phoneNumber: '0987654321' })
        .expect(200);
      profile = await supertest({ userId: user.id }).get('/profile');
      expect(profile.body?.phoneNumber).toEqual('real0987654321');
    });

    it('lets ta change default teams message', async () => {
      const user = await UserFactory.create();
      let profile = await supertest({ userId: user.id }).get('/profile');
      expect(profile.body?.defaultMessage).toEqual(null);
      await supertest({ userId: user.id })
        .patch('/profile')
        .send({ defaultMessage: "Hello! It's me :D" })
        .expect(200);
      profile = await supertest({ userId: user.id }).get('/profile');
      expect(profile.body?.defaultMessage).toEqual("Hello! It's me :D");
    });

    it('lets ta change includeDefaultMessage', async () => {
      const user = await UserFactory.create();
      let profile = await supertest({ userId: user.id }).get('/profile');
      expect(profile.body?.includeDefaultMessage).toEqual(true);
      await supertest({ userId: user.id })
        .patch('/profile')
        .send({ includeDefaultMessage: false })
        .expect(200);
      profile = await supertest({ userId: user.id }).get('/profile');
      expect(profile.body?.includeDefaultMessage).toEqual(false);
    });

    it('lets user change email', async () => {
      const user = await UserFactory.create();
      let profile = await supertest({ userId: user.id }).get('/profile');
      const newEmail = 'new_test_email@ubc.ca';

      expect(profile.body?.email).toEqual(user.email);
      await supertest({ userId: user.id })
        .patch('/profile')
        .send({ email: newEmail })
        .expect(200);

      profile = await supertest({ userId: user.id }).get('/profile');
      expect(profile.body?.email).toEqual(newEmail);
    });

    it('fails to change user email when email is used by another user', async () => {
      const user = await UserFactory.create();

      await supertest({ userId: user.id })
        .patch('/profile')
        .send({ email: user.email })
        .expect(400);
    });
  });
});
