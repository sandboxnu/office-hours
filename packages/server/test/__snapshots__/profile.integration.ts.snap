// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Integration GET /profile returns the logged-in user profile 1`] = `
Object {
  "courses": Array [
    Object {
      "course": Object {
        "id": 1,
        "name": "CS 2500",
      },
      "role": "student",
    },
  ],
  "defaultMessage": null,
  "desktopNotifs": Array [],
  "desktopNotifsEnabled": false,
  "email": "user@ubc.ca",
  "firstName": "User",
  "id": 1,
  "includeDefaultMessage": true,
  "insights": Array [
    "TotalStudents",
    "TotalQuestionsAsked",
    "MedianWaitTime",
    "QuestionTypeBreakdown",
    "MostActiveStudents",
    "QuestionToStudentRatio",
    "MedianHelpingTime",
  ],
  "lastName": "Person",
  "name": "User Person",
  "organization": Object {
    "id": 1,
    "orgId": 1,
    "organizationDescription": "UBC Okanagan",
    "organizationLogoUrl": null,
    "organizationName": "UBCO",
    "organizationRole": "member",
  },
  "phoneNotifsEnabled": false,
  "photoURL": null,
  "sid": null,
  "userRole": "user",
}
`;
